src/veranda/io/geotiff.py

```diff
@@ -341,135 +311,160 @@ def __init__(self, filename, mode='r', count=None, compression='LZW',
+            self.src = self.driver.Create(self.filepath, n_cols, n_rows,
+                                          n_bands, gdal_dtype[dtype],
                                           self.gdal_opt)
 
-            if self.geotransform:
-                self.src.SetGeoTransform(self.geotransform)
+            if self.geotrans is not None:
+                self.src.SetGeoTransform(self.geotrans)
 
-            if self.spatialref:
-                self.src.SetProjection(self.spatialref)
+            if self.sref is not None:
+                self.src.SetProjection(self.sref)
 
             if self.tags is not None:
                 if 'description' in self.tags:
                     self.src.SetDescription(self.tags['description'])
-                if 'metadata' in self.tags:
-                    self.src.SetMetadata(self.tags['metadata'])
                 if 'gcps' in self.tags:
                     if len(self.tags['gcps']) != 0:
                         self.src.SetGCPs(self.tags['gcps'],
                                          self.tags['spatialreference'])
 
-        if self.mode == 'r':
-            self.src = gdal.Open(self.filename, gdal.GA_ReadOnly)
+            if self.metadata is not None:
+                self.src.SetMetadata(self.metadata)
 
-            self.n_layers = self.src.RasterCount
+            if n_bands == 1:
+                self.shape = (n_rows, n_cols)
+            else:
+                self.shape = (n_bands, n_rows, n_cols)
+
+        if self.mode == 'r':
+            self.src = gdal.Open(self.filepath, gdal.GA_ReadOnly)
 
-            if self.n_layers == 1:
+            n_rasters = self.src.RasterCount
+            if n_rasters == 1:
                 self.shape = (self.src.RasterYSize, self.src.RasterXSize)
             else:
-                self.shape = (self.src.RasterCount, self.src.RasterYSize,
-                              self.src.RasterXSize)
+                self.shape = (n_rasters, self.src.RasterYSize, self.src.RasterXSize)
 
-            self.ndim = len(self.shape)
-            self.geotransform = self.src.GetGeoTransform()
-            self.spatialref = self.src.GetProjection()
+            self.geotrans = self.src.GetGeoTransform()
+            self.sref = self.src.GetProjection()
+
+            self.metadata = self.src.GetMetadata()
 
             # assume all bands are like the first
-            self.dtype = gdal.GetDataTypeName(
-                self.src.GetRasterBand(1).DataType)
+            self.dtype = gdal.GetDataTypeName(self.src.GetRasterBand(1).DataType)
 
             self.chunks = list(self.src.GetRasterBand(1).GetBlockSize())[::-1]
 
         if self.src is None:
-            raise IOError("Open failed: %s".format(self.filename))
+            raise IOError("Open failed: %s".format(self.filepath))
 
-    def read(self, band=1, return_tags=True, encode_func=None, sub_rect=None):
+    def read(self, row=None, col=None, n_rows=1, n_cols=1, band=None, nodataval=None,
```

Comment:
I'm not sure about row, col and n_rows, n_cols? What are the benefits to split it in this way?

Reply:
That it is in alignment with all of our future interfaces, starting with a raster geometry. Maybe `row`, `col`, `height` and `width` is better.